# Explore data

In [45]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import datetime

%matplotlib inline

In [46]:
network = pd.read_csv("BI_intern_homework_dataset1.csv")
print(network.head())

         date source_game target_game country  ad_impressions  installs
0  2020-07-01           A           K      US           32105      1380
1  2020-07-02           A           K      US           29535      1430
2  2020-07-03           A           K      US           28702      1262
3  2020-07-04           A           K      US           32822      1550
4  2020-07-05           A           K      US           23097      1034


In [47]:
print(network.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15500 entries, 0 to 15499
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   date            15500 non-null  object
 1   source_game     15500 non-null  object
 2   target_game     15500 non-null  object
 3   country         15500 non-null  object
 4   ad_impressions  15500 non-null  int64 
 5   installs        15500 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 726.7+ KB
None


The dataset 1 is already clean for further analysis.

# Data Analysis

## 1.1 Which source game has shown the most ad impressions?

In [48]:
# group the network dataframe by source games
ads_by_source_game = network.groupby("source_game")

# find the total ad impressions for each source game
# and sort from the largest to smallest
ads_by_source_game.ad_impressions.sum().sort_values(ascending=False)

source_game
J    39561455
F    35873352
I    34218532
A    33415518
D    32558451
G    28989621
H    24411922
C    23422934
B    12882970
E    12763510
Name: ad_impressions, dtype: int64

**CONCLUSION: It was shown in the table above that source game J has shown the most ad impressions with a total of 39,561,455 ads.** 

## 1.2 Which target game has gotten the most installs?

In [49]:
# group the network dataframe by target games
installs_by_target_game = network.groupby("target_game")

# find the total installs for each target game
# and sort from the largest to smallest
installs_by_target_game.installs.sum().sort_values(ascending=False)

target_game
T    903191
R    609869
L    547293
O    533252
Q    471479
M    459650
N    444215
K    418441
S    406413
P    360226
Name: installs, dtype: int64

The same process is conducted to find the target game having the most installations. **It shows that target game T has the most installations with a total of 903,191 installs.**

## 1.3 On July 4, there is an increase in the number of installs. Find drivers for this change in the data set

In [50]:
# total installs for each day in July
network.groupby("date").installs.sum().sort_values(ascending=False)

date
2020-07-04    222203
2020-07-12    189985
2020-07-05    188568
2020-07-03    183057
2020-07-02    180930
2020-07-06    179042
2020-07-26    178167
2020-07-16    175164
2020-07-19    171296
2020-07-11    171247
2020-07-01    170227
2020-07-25    167683
2020-07-24    167471
2020-07-07    166369
2020-07-08    165875
2020-07-27    163672
2020-07-13    162743
2020-07-18    162594
2020-07-15    162580
2020-07-10    161130
2020-07-17    160404
2020-07-20    160027
2020-07-23    159303
2020-07-09    155501
2020-07-28    151069
2020-07-14    150658
2020-07-29    150134
2020-07-21    149938
2020-07-22    145128
2020-07-30    144447
2020-07-31    137417
Name: installs, dtype: int64

In [51]:
# average installs for a day in July
network.groupby("date").installs.sum().sort_values(ascending=False).mean()

166259.0

In [52]:
# since it is quite hard to see in sum number, I decided to change to mean values
# mean installs for each day in July
network.groupby("date").installs.mean().sort_values(ascending=False)

date
2020-07-04    444.406
2020-07-12    379.970
2020-07-05    377.136
2020-07-03    366.114
2020-07-02    361.860
2020-07-06    358.084
2020-07-26    356.334
2020-07-16    350.328
2020-07-19    342.592
2020-07-11    342.494
2020-07-01    340.454
2020-07-25    335.366
2020-07-24    334.942
2020-07-07    332.738
2020-07-08    331.750
2020-07-27    327.344
2020-07-13    325.486
2020-07-18    325.188
2020-07-15    325.160
2020-07-10    322.260
2020-07-17    320.808
2020-07-20    320.054
2020-07-23    318.606
2020-07-09    311.002
2020-07-28    302.138
2020-07-14    301.316
2020-07-29    300.268
2020-07-21    299.876
2020-07-22    290.256
2020-07-30    288.894
2020-07-31    274.834
Name: installs, dtype: float64

In [53]:
# average installs for a day in July
network.installs.mean()

332.518

It can be seen that on 4th of July there is a peak in installation numbers with 222,203 installs (mean of 444 installs of a game in a day) out of 166,259 installs average for each day (mean of 332 installs). This means an increase of almost 60,000 installs.

In [54]:
# dataframe for 04.07.2020
network_040720 = network[network.date == "2020-07-04"]

# the total installs for each game on 04.07.2020
network_040720.groupby("target_game").installs.sum().sort_values(ascending=False)

target_game
R    43933
T    36637
K    25709
O    22089
L    19998
Q    18866
S    16182
M    16059
P    12394
N    10336
Name: installs, dtype: int64

In [55]:
# dataframe for target game R on 04.07.2020
print(network_040720[network_040720.target_game == "R"])

             date source_game target_game country  ad_impressions  installs
220    2020-07-04           A           R      US           97223       890
530    2020-07-04           B           R      US            8079       423
840    2020-07-04           C           R      US           82929      1509
1150   2020-07-04           D           R      US           29642      1214
1460   2020-07-04           E           R      US           20922       855
1770   2020-07-04           F           R      US            6972       699
2080   2020-07-04           G           R      US            5526       141
2390   2020-07-04           H           R      US            6216       302
2700   2020-07-04           I           R      US          599130     14573
3010   2020-07-04           J           R      US          136547      2927
3320   2020-07-04           A           R      RU           13052       128
3630   2020-07-04           B           R      RU            2755       141
3940   2020-

It seems that the company try to put R game ad impressions in large quantity to source games in all 5 countries (GB: 200,000 thourgh source_game I; BR: 127,547 through source_game I; IN: 40,378 through source_game I and 46,093 through source_game J; US: 599,130 through source_game I and 136,547 through source_game J; RU: 41,337 through source_game I, 43,719 through source_game J and 69,369 through source_game C. This results in larger people seeing the advertisements and consequently, more would install game R.
<br>
<br>
As analysed above, R game is second most installed game of all targets game, while source games I and J lie in the top 3 most ad impression source games. This could be a promotion day for target game R across 5 countries through the most ad impressions source games (I and J). 
T game is also worth scrutinized since it was the most installed target game and its number of installations also increased drastically on 04.07.2020. 

In [56]:
# data frame for target game T on 04.07.2020
print(network_040720[network_040720.target_game == "T"])

             date source_game target_game country  ad_impressions  installs
282    2020-07-04           A           T      US           82713      1304
592    2020-07-04           B           T      US           10317       458
902    2020-07-04           C           T      US          111582      4710
1212   2020-07-04           D           T      US           32061      2438
1522   2020-07-04           E           T      US           13868       584
1832   2020-07-04           F           T      US           21236       945
2142   2020-07-04           G           T      US           12369       620
2452   2020-07-04           H           T      US           12809       647
2762   2020-07-04           I           T      US          122946      3447
3072   2020-07-04           J           T      US          244383      1448
3382   2020-07-04           A           T      RU           11105       176
3692   2020-07-04           B           T      RU            3424       156
4002   2020-

It is suprising that more target games were installed through ads shown in source games C and D, which are not the top ad impressions source games,nevertheless the attempt to promote through source game J.

In [57]:
# data frame of target game K on 04.07.2020
print(network_040720[network_040720.target_game == "K"])

             date source_game target_game country  ad_impressions  installs
3      2020-07-04           A           K      US           32822      1550
313    2020-07-04           B           K      US           13787      1098
623    2020-07-04           C           K      US            9464       774
933    2020-07-04           D           K      US           26046       783
1243   2020-07-04           E           K      US           16556       566
1553   2020-07-04           F           K      US            3321       144
1863   2020-07-04           G           K      US          168861      3987
2173   2020-07-04           H           K      US           18044       731
2483   2020-07-04           I           K      US           47916       669
2793   2020-07-04           J           K      US           37764      1340
3103   2020-07-04           A           K      RU            4280       218
3413   2020-07-04           B           K      RU            4393       371
3723   2020-

The same trend for target game R appears also to target game K with a vast quantity of ads shown through source game G all across 5 countries. This results in a larger number of installations. This is quite interesting since target game K is not the favoured game to installs, at least in the chart of most installed target game with position 7 out of 10.  
<br>
<br>
**CONCLUSION: The increase in the number of installs may be attributed to the promotion (more ad impressions) all across 5 countries of target game R though source games I and J, target game K through source game G and target game T through the same source games but the results show more game installs in source games in C and D. This results in more people seeing these target games while being in source games, and leading to more installs. Especially there were almost 600,000 target game R's ads shown in source game I in the US, which results in a peak of around 14,500 installs, the highest number of installs in the month.**

In [58]:
K_G = network_040720[(network_040720.source_game=="G") 
                     & (network_040720.target_game=="K")]
R_I_J = network_040720[((network_040720.source_game=="I") 
                        | (network_040720.source_game=="J"))
                       & (network_040720.target_game=="R")]
T_C_D = network_040720[((network_040720.source_game=="C") 
                        | (network_040720.source_game=="D"))
                       & (network_040720.target_game=="T")]

# the highest number of installs in July
print(network.installs.max())

# the final table for drivers for increase in installs
conclusion = pd.concat([K_G, R_I_J, T_C_D])
print(conclusion)

14573
             date source_game target_game country  ad_impressions  installs
1863   2020-07-04           G           K      US          168861      3987
4963   2020-07-04           G           K      RU           38538       866
8063   2020-07-04           G           K      IN           73903      1745
11163  2020-07-04           G           K      BR           75318      1783
14263  2020-07-04           G           K      GB           80757      1649
2700   2020-07-04           I           R      US          599130     14573
3010   2020-07-04           J           R      US          136547      2927
5800   2020-07-04           I           R      RU           41337       911
6110   2020-07-04           J           R      RU           43719       976
8900   2020-07-04           I           R      IN           40378       997
9210   2020-07-04           J           R      IN           46093       958
12000  2020-07-04           I           R      BR          127547      2998
12310 

## 2.1 In data set 1, there's an increase in average conversion rate across all target gameson July 9-10 even though there isn't a clear increase in the conversion rate of any of the target games. Explain this phenomenon.

In [59]:
network_090720 = network[network.date =="2020-07-09"]
network_100720 = network[network.date == "2020-07-10"]

In [60]:
print(network_090720.installs.sum())

155501


In [61]:
print(network_100720.installs.sum())

161130


In [62]:
print(network_090720.ad_impressions.sum())

9144758


In [63]:
print(network_100720.ad_impressions.sum())

8393952


In [64]:
# conversion rate on 09.07.2020
a = network_090720.installs.sum()/network_090720.ad_impressions.sum()*100
print(a)

1.7004386556757436


In [65]:
# conversion rate on 10.07.2020
b = network_100720.installs.sum()/network_100720.ad_impressions.sum()*100
print(b)

1.9195963951187711


As shown above, the calculated conversion rates illustrate average conversion rate of date 09.07.2020 and 10.07.2020. The difference in these 2 days lie in the fact that on 9th of July, there were less installs but more ad impressions compared on 10th of July, which results in the conversion rate of 9th July is less than 10th of July.
For instance: 1/3 < 2/2 (1<2, 3>2)
Given the fact that conversion rate is proportionally to success rate of people actually installing target games. These calculated conversion rates shows that the perforamce on date 10.07.2020 is better than date 09.07.2020 ( with more installs and less ad impressions).

In [36]:
network_090720.groupby("target_game").installs.sum()/network_090720.groupby("target_game").ad_impressions.sum()*100

target_game
K    3.741771
L    0.730085
M    2.419663
N    2.243864
O    2.575083
P    2.275052
Q    2.127058
R    2.069750
S    0.862059
T    2.522888
dtype: float64

In [37]:
network_100720.groupby("target_game").installs.sum()/network_100720.groupby("target_game").ad_impressions.sum()*100

target_game
K    3.586167
L    0.697190
M    2.194735
N    2.252687
O    2.583943
P    2.124973
Q    2.063748
R    1.909626
S    0.827991
T    2.488650
dtype: float64

In [73]:
print(network_090720.groupby("target_game").installs.sum())

target_game
K    13288
L    19207
M    15786
N     7548
O    16356
P     8430
Q    15120
R    17269
S    12240
T    30257
Name: installs, dtype: int64


In [71]:
print(network_100720.groupby("target_game").installs.sum())

target_game
K    13678
L     8534
M    15311
N     7557
O    16500
P     9172
Q    35546
R    18391
S     6644
T    29797
Name: installs, dtype: int64


In [76]:
print(network_090720.groupby("target_game").ad_impressions.sum())

target_game
K     355126
L    2630789
M     652405
N     336384
O     635164
P     370541
Q     710841
R     834352
S    1419856
T    1199300
Name: ad_impressions, dtype: int64


In [72]:
print(network_100720.groupby("target_game").ad_impressions.sum())

target_game
K     381410
L    1224056
M     697624
N     335466
O     638559
P     431629
Q    1722400
R     963068
S     802424
T    1197316
Name: ad_impressions, dtype: int64


However, if we take each of conversion rate for each game on 10.07.2020, the results actually show that they almost perform worse than on 09.07.2020 and for some games, the rate almost remained the same. 
Moreover, if we take a look at the total ad impressions for each game, we can see that the total amount was dropped almost by half for 2 lowest installed target games (almost 2 millions ads) and then the amount of ad impressions was increased for those games having above two point percentage of conversion rate, with around 75,000 drop in total ad impressions on 10th of July. 
<br>
<br>
If we do calculation, if we drop ad impressions for these 2 lowest installed target game L, Q to have the same amount of ads on 10.07.20 and they still have the same conversion rate, 0.73 and 0.86 percentage respectively, there were only 10,270 and 5,322 less installed that is about 15,000 installs dropped. But then we have 20,426 installed games increased for only 1 million ad impressions increase on 10.07.2020, which is already 5,000 increase for installations compared with 15,000 installs dropped. And slightly change the ad impressions of the other games to gain higher number of installs, resulting a greater total of installations per day. Since the ad impressions amount was dropped from 2 lowest conversion rate games for over 2 millions and increase 1 million for better conversion rate game, and distribute slighly more or less ad impressions for other games, the ad impressions decrease and total installations increase, which explains why the average conversion rate increase from 09.07 to 10.07. But since the conversion rate will not be the same daily, increasing the amount of ad impressions will not promise to have the proportional increase in installs, but at least we know that they increase and vice versa. And since the installs increase and ad_impression increase quite proportionally, the conversion rate stays  
<br>
<br>
**CONCLUSION: This means that in graph of conversion rate, the conversion rates seem not to have a significant change, but in fact, there has been a great modification of ad impressions and installations for each game on 10.07.2020, leading to better performance of average conversion rate (0.2 % increase). This might be conducted by some algorithms to maximize and predict the installs of each target game from 09.07.2020 to 10.07.2020.**
<br>
<br>
## 3.1 Calculate the average conversion rate of the target game

In [86]:
target = pd.read_csv("BI_intern_homework_dataset2.csv")
print(target.info())

print(target.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 724 entries, 0 to 723
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         724 non-null    object
 1   target_game  724 non-null    object
 2   source_game  724 non-null    object
 3   campaign     724 non-null    object
 4   event_type   724 non-null    object
dtypes: object(5)
memory usage: 28.4+ KB
None
         date target_game source_game campaign     event_type
0  2020-07-01           U           Z        W  ad_impression
1  2020-07-01           U           Y        V  ad_impression
2  2020-07-01           U           Y        W  ad_impression
3  2020-07-01           U           Z        W        install
4  2020-07-01           U           Z        W  ad_impression


In [100]:
a = target[target.event_type == "install"]
b = target[target.event_type == "ad_impression"]

conversion_rate = len(a)/len(b)*100
print(conversion_rate)

13.479623824451412


**CONLCUSION: The average conversion rate of the target game is 13.5%**